File: bench/cpp/sum.cpp
// Control: see bench/goose/sum.goose.
//   VARIANT 0: vector, no reserve.
//   VARIANT 1: vector with an exact reserve.
#include "bench.h"
#include <vector>

int main() {
    std::vector<int32_t> data;
#if VARIANT == 1
    data.reserve(BENCH_N);
#endif
    uint64_t r = 12345;
    for (long long i = 0; i < BENCH_N; i++) {
        r = xs_next(r);
        data.push_back((int32_t)xs_mod(r, 1000));
    }
    long long total = 0;
    for (long long p = 0; p < 8; p++)
        for (int32_t x : data) total += x ^ p;
    emit((long long)data.size());
    emit(total);
    return 0;
}

File: bench/rust/sum.rs
// Control: see bench/goose/sum.goose. Collecting a mapped range is the way a
// Rust programmer writes "build a vector of N values", and because `Range` is
// `TrustedLen` the allocation is exact -- so Rust gets the C++ `reserve` row's
// behaviour from the idiomatic spelling, with nothing to remember.
mod bench;
use bench::*;

const N: i64 = 1000000;                  // BENCH_N
const PASSES: i64 = 8;

fn main() {
    let mut r: u64 = 12345;
    let data: Vec<i32> = (0..N).map(|_| { r = xs_next(r); xs_mod(r, 1000) as i32 }).collect();
    let mut total: i64 = 0;
    for p in 0..PASSES {
        // The `^ p` keeps the passes from collapsing into one multiply.
        for &x in &data { total += (x as i64) ^ p; }
    }
    emit(data.len() as i64);
    emit(total);
}
